HPCC-29549 Handle fully qualified path when validating DZ scope access

wangkx opened the pull request:
1. Move the code of the scope permission validation from the checkDZScopeAccessAndSetSpraySourceDFUFileSpec() to the readAndCheckSpraySourceReq().
2. In readAndCheckSpraySourceReq(), parse the file requests inside the SourcePath request. For each file request, validating the scope permission. If a file request contains a host/ip, validate the host/ip and local file path. If a file request does not contain a host/ip, validate the file path.
3. Change the checkDZScopeAccessAndSetSpraySourceDFUFileSpec() to setSpraySourceDFUFileSpec().
4. Do not validate DZ scope if no DZ configured and no DropZoneRestriction.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-29549 Handle fully qualified path when validating DZ scope access

1. Move the code of the scope permission validation from the
checkDZScopeAccessAndSetSpraySourceDFUFileSpec() to the
readAndCheckSpraySourceReq().
2. In readAndCheckSpraySourceReq(), parse the file requests inside
the SourcePath request. For each file request, validating the scope
permission. If a file request contains a host/ip, validate the
host/ip and local file path. If a file request does not contain a
host/ip, validate the file path.
3. Change the checkDZScopeAccessAndSetSpraySourceDFUFileSpec() to
setSpraySourceDFUFileSpec().

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
Noticed 1 smoketest failure: Spray variable with wrong queue name: Fail. But, I do not know why the queue name is wrong.

wangkx commented:
@jakesmith a smoketest failed. I will take a look.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
-                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Source network IP not specified.");
-            Owned<IPropertyTree> plane = getAndValidateDropZone(sourcePath, sourceIPReq);
-            sourcePlaneReq.append(plane->queryProp("@name"));
+            if (firstFileReq)
```
could this be if (nullptr == dropZone) ?

And, it should be an error in cloud and in bare-metal unless isDropZoneRestrictionEnabled() = false if the dropzone cannot be found.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+                firstFileReq = false;
+
+                //Use the fileReq and sourceHostReq to find out dropzone plane.
+                dropZone.setown(getAndValidateDropZone(fileReq, sourceHostReq));
```
it feels this should proceed the if (isAbsolutePath(fileReq)) block (code between lines 1965-1978).
It also might make this method clear, if that code were in a separate (static) helper function (can still be within ws_fsService.cpp).

wangkx replied:
If this proceeds the if (isAbsolutePath(fileReq)) block (code between lines 1965-1978), the getAndValidateDropZone() will be called when the Source Plan is not empty.
I am not sure we want to call the getAndValidateDropZone() when the Source Plane is not empty. When the Source Plane is not empty, the sourceHostReq has been validated and the fileReq will be validated inside the  if (isAbsolutePath(fileReq)) block.

jakesmith reviewed (commented):
@wangkx - I think the consolidation into readAndCheckSpraySourceReq and checking of the UNC paths is ok, but the code is quite hard to follow at the moment, and I am not sure the order things are validated is correct.
I think it could be re-organised and made clearer. See comments.

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+    StringBuffer sourcePath, s(srcPath);
+    s.trim();
+    if (s.isEmpty())
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Source path not specified.");
```
Will change

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
-    sourcePath.trim();
-    if(srcxml.length() == 0)
+    if(srcxml.length() > 0)
+        return;
```
Will change

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,132 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+        {
+            splitUNCFilename(fileReq, &hostInPathStr, &localPath, &localPath, &localPath);
+            const char* hostInPath = hostInPathStr.str() + 2;
+            if (!isEmptyString(srcPlane))
```
Will change

wangkx commented:
> the code is quite hard to follow at the moment,
I am going to re-organize the code to see if it will be better.

wangkx committed:
HPCC-29549 Revise based on code review

1. Change several throw exception calls.
2. Check the srcxml in the outside of the readAndCheckSpraySourceReq().
3. Validate the hostInPath with the srcHost if the srcHost is supplied.
4. Re-organize the code.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith I did the changes and re-organized the code. Please review.

jakesmith commented:
@wangkx - PR and commit message should be "Handle fully qualified .. "

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+    if (!sourcePlaneReq.isEmpty())
+        getStandardPosixPath(sourcePath, s.str());
+    else
+        getStandardPosixPath(sourcePathReq, s.str());
```
sorry, I don't understand the logic here..
If no plane in request, then fill incoming srcPath, into [outbound] sourcePathReq, leaving 'sourcePath' empty,
which is then used below (line 1935)..

Needs comments to clarify at least.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+        return;
+
+    splitUNCFilename(fileReq, &hostInPathStr, &localPath, &localPath, &localPath);
+    const char* hostInPath = hostInPathStr.str() + 2;
```
needs/using bit after '//', i.e. the <hostname>, but returning (in hostInPathStr) the //<hostname> ..
that is confusing.
Should return the hostname - that if anything will be whats needed, caller should have to do ... + 2

Code clarity here is far more important than a slight gain in efficiency because need to copy in to output StringBuffer.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+        checkHostInSpraySourcePath(fileReq, sourcePlaneReq, dropZone, sourceHostReq, &sourceHostEp, localPath, hostInPath);
+
+        //Validate file path.
+        const char* path = localPath.isEmpty() ? fileReq : localPath.str();
```
would be clearer if checkHostInSpraySourcePath had returned true if was a UNC path and had validated, and do, e.g.:
```
const char *path;
if (IsHostInSprayPath(fileReq, sourcePlaneReq, dropZone, sourceHostReq, &sourceHostEp, localPath, hostInPath))
    path = localPath.str();
else
   path = fileReq;

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+    StringArray fileReqs;
+    fileReqs.appendList(sourcePath, ","); // handles comma separated files
+
+    if (sourcePlaneReq.isEmpty())
```
let us try to separate and label code like this as much as possible to emphasis it's legacy.
A comment here and above the definition of 'checkSpraySourcePathAndGetPlane' will suffice.

It will make future reading easier to follow the standard/normal/expected code path.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2008,7 +2118,7 @@ bool CFileSprayEx::onSprayFixed(IEspContext &context, IEspSprayFixed &req, IEspS
 
         IDFUfileSpec *source = wu->queryUpdateSource();
-        checkDZScopeAccessAndSetSpraySourceDFUFileSpec(context, sourcePlaneReq, sourceIPReq, srcfile, srcxml, source);
+        setSpraySourceDFUFileSpec(sourceHostReq, sourcePathReq, srcxml, source);
```
This is probably going to need revising to ensure it includes plane name, following this change: https://github.com/hpcc-systems/HPCC-Platform/pull/17353/files#diff-8b47ea38bf0ea671d5db7b9ae0166a6c19a1ad525011ffac7556726d8245844a

And for that reason and others this PR should be rebased onto master.

jakesmith replied:
>Not sure why we need to 'ensure it includes plane name'? Do we want to call the source->setGroupName(sourcePlaneReq); inside the setSpraySourceDFUFileSpec()?

See https://track.hpccsystems.com/browse/HPCC-29553
It ensures the group names are created (in bare-metal) for planes, meaning the logical files (IFileDescriptor's) will have properly formed group not IP's within them.
The IFileDescriptor's created by DFU need to know the group(plane) name, and hence this (setGroupName) is the means by which they do.

jakesmith replied:
>If this PR is rebased onto master, should we create another small 9.0.x PR to fix the crash when isDropZoneRestrictionEnabled = false and getAndValidateDropZone() returns nullptr and to not call getDZFileScopePermissions() if no plane name is found?

Yes, that makes sense.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+
+        if (firstFileReq)
+        {
+            firstFileReq = false;
```
still seems to a logical problem here..
What happens if 2nd, 3rd UNC path have different hostname in them?

Looks like 1st will be validated, and all others ignored at the moment?
But judging by setSpraySourceDFUFileSpec and srcDFUfileSpec->setMultiFilename .. we don't support it anyway.. i.e. only 1 host has ever been permitted here.

If so, this should not ignore other UNC with different hosts, but throw an error and reject it

wangkx replied:
I do assume all of the UNC paths have the same hostname now. If there are different hostnames, could they are in different planes? If so, which one should be set into the DFU WU or we throw an error?

jakesmith replied:
as I say, afaics, srcDFUfileSpec->setMultiFilename etc. wouldn't support it anyway, so should throw an error.

wangkx replied:
I just realized that the checkHostInSpraySourcePath() does throw an error if hostnames are from different hosts. If a dropzone is found for the first file, the dropzone will be used for all files.

wangkx replied:
The code should be revised to check the path also if a dropzone is found: if (!startsWith(path, dropZonePlanePath)) ...

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2133,9 +2243,9 @@ bool CFileSprayEx::onSprayVariable(IEspContext &context, IEspSprayVariable &req,
-        const char* srcfile = sourcePathReq.str();
+        StringBuffer sourcePlaneReq, sourceHostReq, sourcePathReq;
+        if (srcxml.length() == 0)
+            readAndCheckSpraySourceReq(context, req.getSourceIP(), req.getSourcePath(), req.getSourcePlane(), sourcePlaneReq, sourceHostReq, sourcePathReq);
```
not for now: there's a lot in common with onSprayFixed vs onSprayVariable, but differences that I can't immediately see why are there, e.g. onSprayVariable has:
```
        StringBuffer gName, ipAddr;
        const char *pTr = strchr(destNodeGroup, ' ');
        if (pTr)
        {
            gName.append(pTr - destNodeGroup, destNodeGroup);
            ipAddr.append(pTr+1);
        }
        else
            gName.append(destNodeGroup);
```

They should be as similar as possible, preferably using common code wherever possible.
Not for now, but could the ecm EspSpray*'s be commoned up and the *Fixed vs *Variable definitions be derived from EspSprayCommon? Such that these methods wouldn't have to duplicate as much?

jakesmith reviewed (commented):
@wangkx - I think it's clearer, but still quite quite convoluted and hard to follow the 'normal' code path vs the edge/exception/legacy cases.
To some extent this is unavoidable because we are having to be backward compatible and support weird cases.
Please see comments.

wangkx commented:
Will add more changes after https://github.com/hpcc-systems/HPCC-Platform/pull/17378.

wangkx commented:
I am closing this PR for now. Will reopen after https://github.com/hpcc-systems/HPCC-Platform/pull/17378.
